roofit/roofitcore/src/RooAbsData.cxx

```diff
@@ -2168,10 +2168,7 @@ void RooAbsData::optimizeReadingWithCaching(RooAbsArg& arg, const RooArgSet& cac
     // Go over all used observables and check if any of them have parameterized
     // ranges in terms of pruned observables. If so, remove those observable
     // from the pruning list
-    TIterator* uIter = usedObs->createIterator() ;
-    RooAbsArg* obs ;
-    while((obs=(RooAbsArg*)uIter->Next())) {
-      RooRealVar* rrv = dynamic_cast<RooRealVar*>(obs) ;
+    for(auto const* rrv : dynamic_range_cast<RooRealVar*>(*usedObs)) {
       if (rrv && !rrv->getBinning().isShareable()) {
    RooArgSet depObs ;
```

Comment:
Maybe fix indentation in 2nd commit if you touch this loop.

Reply:
> This looks quite good already, but consider removing the helper class.
> 
> ```c++
> if (templateParameter)
> ```
> 
> won't have any runtime overhead with any decent compiler, and both `dynamic_cast` and `static_cast` are syntactically correct code paths, so no `if constexpr` required.

Thanks for the comments! Before I address the other comments: just one answer to this comment. No, it seems that `dynamic_cast<T*>` is not syntactically correct if T is not polymorphic, e.g. it has not virtual method.

You can for example see that this code doesn't compile:

```C++
struct A {
    void foo() {}
};

struct B : public A {
    void bar() {}
};

int main() {

    A* ptr = new B;
    dynamic_cast<B*>(ptr);
    delete ptr;
}
```
Here is the error:
```
collection_cast.cc: In function ‘int main()’:
collection_cast.cc:95:5: error: cannot ‘dynamic_cast’ ‘ptr’ (of type ‘struct A*’) to type ‘struct B*’ (source type is not polymorphic)
   95 |     dynamic_cast<B*>(ptr);
      | 
```

So we have to find a way to make the decision static_cast/dynamic_cast at compile time.